Feedback 형식 변경

kochungcheon opened the pull request:
구현 한 것
- feedback 형식 변경
- llm feedback 실패 시 (json parse 실패나 feedback 저장 실패) llm call retry 3번
- pending true 일 때, 일기 요청 할 수 없는 로직 개선
- feedback key에 . 혹은 $ 가 오면 빈 값으로 대체

kochungcheon committed:
refactor : folder 구조 수정
- remove com

kochungcheon committed:
refactor : llm 피드백 요청 형식 변경

kochungcheon committed:
test : llm 피드백

kochungcheon committed:
feat : json parse 실패 시 최대 3번 llm요청한다

kochungcheon committed:
refactor : diary pending init value

kochungcheon committed:
refactor : improvedContent, feedback 형식으로 저장

kochungcheon committed:
refactor : pending true 일 때는 일기 피드백을 요청할 수 없다.
- 피드백은 한 번만 요청 가능

kochungcheon committed:
test : Json parse 실패해서 피드백 생성에 실패한다

kochungcheon committed:
feat : feedback key에는 . 혹은 $ 가 오면 빈 값으로 대체
- MongoDB 제약 조건 고려

kochungcheon committed:
test : feedback key에는 . 혹은 $ 가 오면 빈 값으로 대체

kochungcheon committed:
merge: resolve conflict

pear96 commented on diary-llm/src/main/java/chungha/diaryllm/service/LlmApiService.java:
```diff
@@ -43,22 +44,30 @@ public Mono<Diary> createFeedBackAndSave(FeedbackReq req) {
 			.switchIfEmpty(Mono.error(new RuntimeException("일기를 찾을 수 없거나 권한이 없습니다.")))
 			.flatMap(diary -> {
-				if (!Objects.isNull(diary.getImprovedContent())) {
+				if (diary.getImprovedContent() != null) {
```
`if (!Objects.isNull(diary.getImprovedContent()))` 에서 이렇게 바꾼 이유가 뭔가요?

kochungcheon replied:
가독성 때문에 변경했습니다! https://github.com/zakirullin/cognitive-load

pear96 commented on diary-llm/src/main/java/chungha/diaryllm/service/LlmApiService.java:
```diff
@@ -43,22 +44,30 @@ public Mono<Diary> createFeedBackAndSave(FeedbackReq req) {
-		return Mono.fromCallable(() -> openAiChatModel.call(prompt)).subscribeOn(Schedulers.boundedElastic());
+		return Mono.fromCallable(() -> openAiChatModel.call(prompt)).subscribeOn(Schedulers.boundedElastic())
+			.retryWhen(
+				Retry.max(3)
```
`Retry.backoff(시도 횟수, 재시도 시간 간격)` 이라는 메서드도 있다네요.
GPT는 이걸 추천해주긴 하는데, 그렇다고 꼭 backoff를 써야만하는 불가피한것 까진 아닌 것 같구요
그냥 뭐가 더 좋을지 어떻게 생각하나 궁금합니다.

kochungcheon replied:
단순하게 3번 요청하는 거 보다, 개발자가 원하는 대로 수정할 수 있는 장점이 있군요 그리고 OpenAI rate limit을 고려하면 더더욱 그렇고요!

pear96 commented on diary-llm/src/test/java/chungha/diaryllm/service/LlmApiServiceTest.java:
```diff
@@ -155,4 +150,35 @@ void createFeedbackAndSave_success() throws JsonProcessingException {
+		when(llmApiRepository.reserveDiaryUpdate(diaryId, userId)).thenReturn(Mono.just(diary));
+
+		// openAiChatModel.call 3번 모두 실패
+		when(openAiChatModel.call(anyString())).thenThrow(new RuntimeException("LLM error"));
```
오 값을 `anyString()` 이렇게 넣는구나

pear96 commented on diary-llm/src/test/java/chungha/diaryllm/service/LlmApiServiceTest.java:
```diff
@@ -181,4 +182,70 @@ void llmRetryAndPendingFalseOnFailure() throws Exception {
+			.id(diaryId)
+			.userId(userId)
+			.content(content)
+			.pending(false)
```
혹시 pending은 `true`여야 하는거 아닌지 궁금합ㄴㅣ다

kochungcheon replied:
아 맞습니다 true입니다 감사합니다!

pear96 reviewed (commented):
테스트 코드 읽는건 익숙해지지가 않는구만

pear96 commented:
어 나 근데.. 지금 봤는데.. build.gradle에 
`implementation 'org.springframework.boot:spring-boot-starter-web'` 이 아니라
`implementation 'org.springframework.boot:spring-boot-starter-webflux'` 이 들어가야 하는거 아니야?

실행하니까 netty 가 아니라 tomcat으로 실행되고 있어서

kochungcheon committed:
build : webflux 로 수정

ccko-whereable commented:
> 어 나 근데.. 지금 봤는데.. build.gradle에 `implementation 'org.springframework.boot:spring-boot-starter-web'` 이 아니라 `implementation 'org.springframework.boot:spring-boot-starter-webflux'` 이 들어가야 하는거 아니야?
> 
> 실행하니까 netty 가 아니라 tomcat으로 실행되고 있어서

webflux로 변경했습니다 감사합니다!

kochungcheon committed:
fix : pending false -> true
- 테스트 오류 수정

kochungcheon committed:
refactor : backoff 반영하여 로직 개선
- json parse 실패와 api call 실패 시 retry 하게 구현
- 과도하게 많은 재시도 요청이 한번에 외부 api에 가지 않도록 수정

kochungcheon commented:
> 어 나 근데.. 지금 봤는데.. build.gradle에 `implementation 'org.springframework.boot:spring-boot-starter-web'` 이 아니라 `implementation 'org.springframework.boot:spring-boot-starter-webflux'` 이 들어가야 하는거 아니야?
> 
> 실행하니까 netty 가 아니라 tomcat으로 실행되고 있어서

webflux로 변경했습니다 감사합니다!
